Allow '@objc(RuntimeName)' on classes with generic ancestry.

jrose-apple opened the pull request:
This is accomplished by recognizing this specific situation and replacing the `objc` attribute with a hidden `_objcRuntimeName` attribute. This *only* applies to classes that are themselves non-generic (including any enclosing generic context) but that have generic ancestry, and thus cannot be exposed directly to Objective-C.

This commit also eliminates `@NSKeyedArchiverClassName`. It was decided that the distinction between `@NSKeyedArchiverClassName` and `@objc` was too subtle to be worth explaining to developers, and that any case where you'd use `@NSKeyedArchiverClassName` was already a place where the ObjC name wasn't visible at compile time.

This commit does not update diagnostics to reflect this change; we're going to change them anyway.

rdar://problem/32414557

jrose-apple commented:
@swift-ci Please test macOS

jrose-apple commented:
@swift-ci Please smoke test Linux

itaiferber reviewed (approved):
Looks like a cleaner way to handle this. Thanks @jrose-apple!

jrose-apple committed:
Allow '@objc(RuntimeName)' on classes with generic ancestry.

This is accomplished by recognizing this specific situation and
replacing the 'objc' attribute with a hidden '_objcRuntimeName'
attribute. This /only/ applies to classes that are themselves
non-generic (including any enclosing generic context) but that have
generic ancestry, and thus cannot be exposed directly to Objective-C.

This commit also eliminates '@NSKeyedArchiverClassName'. It was
decided that the distinction between '@NSKeyedArchiverClassName' and
'@objc' was too subtle to be worth explaining to developers, and that
any case where you'd use '@NSKeyedArchiverClassName' was already a
place where the ObjC name wasn't visible at compile time.

This commit does not update diagnostics to reflect this change; we're
going to change them anyway.

rdar://problem/32414557

jrose-apple commented:
@swift-ci Please test macOS

jrose-apple commented:
@swift-ci Please smoke test Linux

jrose-apple commented:
non-review cc for @amartini51

jrose-apple commented:
(full tests happening in #10083)

@swift-ci Please smoke test

jrose-apple commented:
Okay, [this one](https://ci.swift.org/job/swift-PR-Linux-smoke-test/8387/)'s not me. @phausler?

> Test Case 'TestURLSession.test_dataTaskWithURLCompletionHandler' started at 2017-06-03 02:21:19.480
TestFoundation/TestNSURLSession.swift:102: error: TestURLSession.test_dataTaskWithURLCompletionHandler : XCTAssertNil failed: "URLError(_nsError: The operation could not be completed)" - error = URLError(_nsError: The operation could not be completed)

@swift-ci Please smoke test Linux

DougGregor commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -1485,6 +1481,12 @@ bool Parser::parseDeclAttribute(DeclAttributes &Attributes, SourceLoc AtLoc) {
+  if (DK == DAK_Count && Tok.getText() == "NSKeyedArchiverClassName") {
+    DK = DAK_ObjC;
+    diagnose(Tok, diag::attr_nskeyedarchiverclassname_removed)
+      .fixItReplace(Tok.getLoc(), "objc");
```
The Fix-It doesn't really work, because `@NSKeyedArchiverClassName` had quotes around the name but `@objc` doesn't, but... I really don't care because "nobody" is using `@NSKeyedArchiverClassName`.

jrose-apple replied:
Yeah. I almost went as far as to not bother at all, especially since we renamed the attributes during development, but still. Fixed in the next commit, as you saw.

DougGregor commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -7022,6 +7022,13 @@ static Optional<ObjCReason> shouldMarkClassAsObjC(TypeChecker &TC,
 
   if (auto attr = CD->getAttrs().getAttribute<ObjCAttr>()) {
     if (kind == ObjCClassKind::ObjCMembers) {
```
After getting confused by this, I feel a deep desire to rename `ObjCClassKind::ObjCMembers` to `ObjCClassKind::ObjCWithGenericAncestry`, especially given that `ObjCMembers` is no longer even true for Swift 4. Anyway, not critical to your change.

DougGregor reviewed (approved):
This looks great. The design is much simpler/cleaner than our previous iteration.

eeckstein commented on lib/IRGen/GenDecl.cpp:
```diff
@@ -948,46 +948,38 @@ void IRGenerator::emitLazyDefinitions() {
-    }
+    assert(CD->getAttrs().hasAttribute<StaticInitializeObjCMetadataAttr>());
+
+    // In this case we don't add a name mapping, but just get the metadata
```
nitpick: The comment needs to be updated, i.e. remove the first part of the sentence, e.g. "// Get the metadata..."

eeckstein reviewed (approved):
LGTM!

jrose-apple commented:
I'm just going to fold this into #10083 at this point.
